src/aiida_quantumespresso/parsers/pw.py

```diff
@@ -106,16 +108,35 @@ def parse(self, **kwargs):
         if exit_code:
             return self.exit(exit_code)
 
-        # If the both stdout and xml exit codes are set, there was a basic problem with both output files and there
-        # is no need to investigate any further.
-        if self.exit_code_stdout and self.exit_code_xml:
-            return self.exit(self.exit_codes.ERROR_OUTPUT_FILES)
+        # If either the stdout or XML were incomplete or corrupt investigate the potential cause
+        if self.exit_code_stdout or self.exit_code_xml:
 
-        if self.exit_code_stdout:
-            return self.exit(self.exit_code_stdout)
+            # First check whether the scheduler already reported an exit code.
+            if self.node.exit_status is not None:
 
-        if self.exit_code_xml:
-            return self.exit(self.exit_code_xml)
+                # The following scheduler errors should correspond to cases where we can simply restart the calculation
+                # and have a chance that the calculation will succeed as the error can be transient.
+                recoverable_scheduler_error = self.node.exit_status in [
+                    PwCalculation.exit_codes.ERROR_SCHEDULER_OUT_OF_WALLTIME.status,
+                    PwCalculation.exit_codes.ERROR_SCHEDULER_NODE_FAILURE.status,
```

Comment:
Where do these exit codes come from, does the scheduler plugin define them or are they actually defined by the calculation plugin?

If the scheduler plugin defines them, can we assume that all scheduler plugins define these exit codes?

Reply:
They are defined by the `CalcJob` base class. However, it is then up to the scheduler plugin to implement the functionality to parse the output and return them if applicable. This is currently done for the `SlurmScheduler` though.